### PR TITLE
feat: initial commit 

### PR DESCRIPTION
**Pull request by samuelOsborne**
This PR provides the start of the base enums for:

- states
- events
- transitions

and the base struct of the state-machine with logic to check which event was received and go to the corresponding state!

### CODE REVIEW

File: dotlottie-sm/src/state_machine.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    pub fn execute_current_state(&mut self, player: &mut DotLottiePlayer) {
+        let current_state = self.current_state.clone();
+        let state_value_result = current_state.write();
+
+        if state_value_result.is_ok() {
+            state_value_result.unwrap().execute(player);
+        }
+    }
+
+    pub fn post_event(&mut self, event: &Event) {
```
Comment (theashraf) [inline dotlottie-sm/src/state_machine.rs:59]: [suggestion]

not important, maybe `emit_event`, or `trigger_event`, `fire_event` ?

**Review by theashraf — COMMENTED** [review]
1. what's the purpose of dotlottie-sm crate ? is it to create or to interpret and run the dotlottie sm ? 

IMO if it's to create then it should live in the dotlottie-fms, if it's to interpret and run, then it should live in dotlottie-rs crate since it's part of the player.

2. Maybe we can crate an independent example crate for the state machines related experiment and keep the demo-player for now ? 
suggested folder structure
   - examples
      - demo-player
      - demo-state-machine
      - demo-x (where x is any feature we want to demo and show)

**Comment by samuelOsborne** [comment]
> 1. what's the purpose of dotlottie-sm crate ? is it to create or to interpret and run the dotlottie sm ?
> 
> IMO if it's to create then it should live in the dotlottie-fms, if it's to interpret and run, then it should live in dotlottie-rs crate since it's part of the player.
> 
> 2. Maybe we can crate an independent example crate for the state machines related experiment and keep the demo-player for now ?
>    suggested folder structure
>    
>    * examples
>      
>      * demo-player
>      * demo-state-machine
>      * demo-x (where x is any feature we want to demo and show)

It was to both create and run state machines (working on the json parser in a separate pr), I thought it might be best as a separate crate as not everyone will need it, and maybe this could help with future efforts of splitting out the player to keep only the crates we need to use?

File: dotlottie-sm/src/state_machine.rs (hunk 1 above)
Comment (samuelOsborne) [reply dotlottie-sm/src/state_machine.rs:59]: this is for the users to post and event to the state machine, I feel like those names feel like they'd be going from the state machine outwards. QT uses `postEvent`

File: dotlottie-sm/src/state_machine.rs [hunk 2]
```diff
@@ -0,0 +1,142 @@
+use std::sync::{Arc, RwLock};
+
+use dotlottie_player_core::DotLottiePlayer;
+
+use crate::{event::Event, state::State, state::StateTrait, transition::TransitionTrait};
+
+pub struct StateMachine {
+    pub states: Vec<Box<State>>,
```
[the remark this answers is not in the document]
Comment (samuelOsborne) [reply dotlottie-sm/src/state_machine.rs:12]: Will remove, currently dont need this vec of states

File: dotlottie-sm/src/state_machine.rs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    pub current_state: Arc<RwLock<State>>,
```
[the remark this answers is not in the document]
Comment (samuelOsborne) [reply dotlottie-sm/src/state_machine.rs:13]: will remove the mutable refs on the methods. 

Im use arc + rwlock on the current_state as I want it to be act as a pointer to the state contained inside a Transition, the Transition will keep ownership of it

File: dotlottie-sm/src/events/event.rs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    fn value(&self) -> bool;
+}
+
+pub trait StringValue {
+    fn value(&self) -> &str;
+}
+
+pub trait NumericValue {
+    fn value(&self) -> f32;
+}
```
[the remark this answers is not in the document]
Comment (samuelOsborne) [reply dotlottie-sm/src/events/event.rs:16]: I'll give this a try, the traits were to be able to get the correct type of the Event

**Comment by samuelOsborne** [comment]
@theashraf Ive added an examples folder, modified the makefile and implemented the comments about shortening the enums

**Pull request merged**